lib/tests.js

```diff
@@ -0,0 +1,405 @@
+
+                multipleListeners((multiErr) => {
+                    instance.removeAllListeners();
+                    instance.disconnect();
+                    callback(multiErr);
+                });
+            });
+        });
+    }
+
+
+    /**
+     * This tests the deduplication functionality. It ensures if a duplicate
+     * message is sent, it is not inserted and requeued is set to true.
+     * It also tests the requeued process. The "requeued" column exists because
+     * we want to run a job one more time but only if the message is being
+     * processed.
+     *
+     * @param {Function} callback
+     */
+    function deduplicationTest(callback) {
+        var emitted;
+
+        emitted = false;
+        instance.on("error", (err) => {
+            instance.removeAllListeners();
+            callback(err);
+        });
+        instance.connect((err) => {
+            if (err) {
+                instance.removeAllListeners();
+                callback(err);
+
+                return;
+            }
+
+            console.log("Sending test message.");
+            instance.sendMessage("message", config.topics, (testErr) => {
+                if (testErr) {
+                    callback(testErr);
+
+                    return;
+                }
+
+                console.log("Test message inserted successfully");
+            });
+            console.log("Testing duplicate message");
+            instance.on("data", (data) => {
+                instance.sendMessage("message", config.topics, (fail) => {
+                    if (fail && !fail.message === "Could not send message.") {
+                        callback(new Error("Was able to insert a duplicate message!"));
+                    } else {
+                        console.log("Duplicate message not inserted, which is expected.");
+                        data.remove((remErr) => {
+                            if (remErr) {
+                                callback(remErr);
+                            }
+
+                            instance.removeAllListeners("data");
+
+                            // Data listener below may get called if the remove above does not work properly.
+                            // Waiting the heartbeatLifetime plus an extra second to ensure it is not emitted.
+                            setTimeout(() => {
+                                if (!emitted) {
+                                    console.log("Requeued message was never emitted again for processing.");
+                                    instance.stopListening();
+                                    instance.removeAllListeners();
+                                    callback(new Error("Duplicate message caused a requeue but message was not emitted for processing a second time."));
+                                }
+                            }, config.heartbeatLifetimeSeconds * 1000 + 1000);
+                            instance.on("data", (moreData) => {
```

Comment:
I think this test is missing one more round of trying to consume the message. We don't know if the message is actually gone at this point right?